<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>ontoProc: Ontology interfaces for Bioconductor, with focus on cell type identification • ontoProc</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="ontoProc: Ontology interfaces for Bioconductor, with focus on cell type identification">
<meta property="og:description" content="ontoProc">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ontoProc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.13.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/ontoProc.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/vjcitn/ontoProc">
    <span class="fas fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>ontoProc: Ontology interfaces for Bioconductor, with focus on cell type identification</h1>
                        <h4 class="author">Vincent J. Carey, stvjc at channing.harvard.edu</h4>
            
            <h4 class="date">March 15, 2021</h4>
      
      
      <div class="hidden name"><code>ontoProc.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>The ambitions of collaborative single cell biology will only be achieved through the coordinated efforts of many groups, to help clarify cell types and dynamics in an array of functional and environmental contexts. The use of formal ontology in this pursuit is well-motivated and research progress has already been substantial.</p>
<p><span class="citation">Bakken et al. (2017)</span> discuss “strategies for standardized cell type representations based on the data outputs from [high-content flow cytometry and single cell RNA sequencing], including ‘context annotations’ in the form of standardized experiment metadata about the specimen source analyzed and marker genes that serve as the most useful features in machine learning-based cell type classification models.” <span class="citation">Aevermann et al. (2018)</span> describe how the FAIR principles can be implemented using statistical identification of necessary and sufficient conditions for determining cell class membership. They propose that Cell Ontology can be transformed to a broadly usable knowledgebase through the incorporation of accurate marker gene signatures for cell classes.</p>
<p>In this vignette, we review key concepts and tasks required to make progress in the adoption and application of ontological discipline in Bioconductor-oriented data analysis.</p>
<p>We’ll start by setting up some package attachments and ontology objects.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://yihui.org/knitr/">knitr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ontoProc</span><span class="op">)</span>
<span class="va">go</span> <span class="op">=</span> <span class="fu"><a href="../reference/getCellOnto.html">getGeneOnto</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">cl</span> <span class="op">=</span> <span class="fu"><a href="../reference/getCellOnto.html">getCellOnto</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">pr</span> <span class="op">=</span> <span class="fu"><a href="../reference/getCellOnto.html">getPROnto</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ontologyPlot</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocStyle">BiocStyle</a></span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span></code></pre></div>
</div>
<div id="scope-of-package" class="section level1">
<h1 class="hasAnchor">
<a href="#scope-of-package" class="anchor"></a>Scope of package</h1>
<p>The following table describes the resources available with <code>get*</code> commands defined in ontoProc. <!-- use inst/scripts/desc.R to generate all but the 'purpose' which is added by hand --></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="va">packDesc2019</span><span class="op">)</span></code></pre></div>
<table class="table">
<colgroup>
<col width="2%">
<col width="15%">
<col width="26%">
<col width="5%">
<col width="4%">
<col width="5%">
<col width="34%">
<col width="4%">
</colgroup>
<thead><tr class="header">
<th align="right">X</th>
<th align="left">func</th>
<th align="left">purpose</th>
<th align="right">nclass</th>
<th align="right">nprop</th>
<th align="right">nroots</th>
<th align="left">datav</th>
<th align="right">fmtv</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">1</td>
<td align="left">getCellLineOnto</td>
<td align="left">Cell line catalog</td>
<td align="right">41780</td>
<td align="right">6</td>
<td align="right">18</td>
<td align="left">NA</td>
<td align="right">NA</td>
</tr>
<tr class="even">
<td align="right">2</td>
<td align="left">getCellOnto</td>
<td align="left">Cell biology concepts</td>
<td align="right">6708</td>
<td align="right">59</td>
<td align="right">38</td>
<td align="left">releases/2018-07-07</td>
<td align="right">1.2</td>
</tr>
<tr class="odd">
<td align="right">3</td>
<td align="left">getCellosaurusOnto</td>
<td align="left">Cell line concepts</td>
<td align="right">87311</td>
<td align="right">6</td>
<td align="right">87311</td>
<td align="left">23</td>
<td align="right">1.2</td>
</tr>
<tr class="even">
<td align="right">4</td>
<td align="left">getChebiLite</td>
<td align="left">Chemicals of biological interest</td>
<td align="right">108496</td>
<td align="right">6</td>
<td align="right">12</td>
<td align="left">155</td>
<td align="right">1.2</td>
</tr>
<tr class="odd">
<td align="right">5</td>
<td align="left">getChebiOnto</td>
<td align="left"></td>
<td align="right">108496</td>
<td align="right">33</td>
<td align="right">12</td>
<td align="left">155</td>
<td align="right">1.2</td>
</tr>
<tr class="even">
<td align="right">6</td>
<td align="left">getDiseaseOnto</td>
<td align="left">Human disease</td>
<td align="right">11283</td>
<td align="right">24</td>
<td align="right">13</td>
<td align="left">releases/2018-06-29</td>
<td align="right">1.2</td>
</tr>
<tr class="odd">
<td align="right">7</td>
<td align="left">getEFOOnto</td>
<td align="left">Experimental factors</td>
<td align="right">20115</td>
<td align="right">6</td>
<td align="right">36</td>
<td align="left">2.87</td>
<td align="right">1.2</td>
</tr>
<tr class="even">
<td align="right">8</td>
<td align="left">getGeneOnto</td>
<td align="left">Gene ontology</td>
<td align="right">47123</td>
<td align="right">43</td>
<td align="right">10</td>
<td align="left">releases/2018-03-27</td>
<td align="right">1.2</td>
</tr>
<tr class="odd">
<td align="right">9</td>
<td align="left">getHCAOnto</td>
<td align="left">Human cell atlas</td>
<td align="right">11047</td>
<td align="right">6</td>
<td align="right">76</td>
<td align="left">NA</td>
<td align="right">NA</td>
</tr>
<tr class="even">
<td align="right">10</td>
<td align="left">getOncotreeOnto</td>
<td align="left">Tumor relations</td>
<td align="right">1298</td>
<td align="right">15</td>
<td align="right">3</td>
<td align="left">ncit/releases/2017-12-15/ncit-oncotree.ttl</td>
<td align="right">1.2</td>
</tr>
<tr class="odd">
<td align="right">11</td>
<td align="left">getPATOnto</td>
<td align="left">Phenotypes and traits</td>
<td align="right">2670</td>
<td align="right">43</td>
<td align="right">21</td>
<td align="left">releases/2018-11-12</td>
<td align="right">1.2</td>
</tr>
<tr class="even">
<td align="right">12</td>
<td align="left">getPROnto</td>
<td align="left">Protein ontology</td>
<td align="right">315957</td>
<td align="right">6</td>
<td align="right">53</td>
<td align="left">57</td>
<td align="right">1.2</td>
</tr>
<tr class="odd">
<td align="right">13</td>
<td align="left">getUBERON_NE</td>
<td align="left">Anatomy</td>
<td align="right">14937</td>
<td align="right">6</td>
<td align="right">135</td>
<td align="left">releases/2017-09-09</td>
<td align="right">1.2</td>
</tr>
</tbody>
</table>
</div>
<div id="methods" class="section level1">
<h1 class="hasAnchor">
<a href="#methods" class="anchor"></a>Methods</h1>
<div id="conceptual-overview-of-ontology-with-cell-types" class="section level2">
<h2 class="hasAnchor">
<a href="#conceptual-overview-of-ontology-with-cell-types" class="anchor"></a>Conceptual overview of ontology with cell types</h2>
<p><strong>Definitions, semantics.</strong> For concreteness, we provide some definitions and examples. We use <code>ontology</code> to denote the systematic organization of terminology used in a conceptual domain. The <code>Cell Ontology</code> is a graphical data structure with carefully annotated terms as nodes and conventionally defined semantic relationships among terms serving as edges. As an example, <code>lung ciliated cell</code> has URI . This URI includes a fixed-length identifier <code>CL_1000271</code> with unambiguous interpretation wherever it is encountered. There is a chain of relationships from <code>lung ciliated cell</code> up through <code>ciliated cell</code>, then <code>native cell</code>, then <code>cell</code>, each possessing its own URI and related interpretive metadata. The relationship connecting the more precise to the less precise term in this chain is denoted <code>SubclassOf</code>. <code>Ciliated cell</code> is equivalent to a <code>native cell</code> that <code>has plasma membrane part</code> <code>cilium</code>. Semantic characteristics of terms and relationships are used to infer relationships among terms that may not have relations directly specified in available ontologies.</p>
<p><strong>Barriers to broad adoption.</strong> Given the wealth of material available in biological ontologies, it is somewhat surprising that formal annotation is so seldom used in practice. Barriers to more common use of ontology in data annotation include: (i) Non-existence of exact matching between intended term and terms available in ontologies of interest. (ii) The practical problem of decoding ontology identifiers. A GO tag or CL tag is excellent for programming, but it is clumsy to co-locate with the tag the associated natural language term or phrase. (iii) Likelihood of disagreement of suitability of terms for conditions observed at the boundaries of knowledge. To help cope with the first of these problems, Bioconductor’s <code>ontologyProc</code> package includes a function <code>liberalMap</code> which will search an ontology for terms lexically close to some target term or phrase. The second problem can be addressed with more elaborate data structures for variable annotation and programming in R, and the third problem will diminish in importance as the value of ontology adoption becomes manifest in more applications.</p>
<p><strong>Class vs. instance.</strong> It is important to distinguish the practice of designing and maintaining ontologies from the use of ontological class terms to annotate instances of the concepts. The combination of an ontology and a set of annotated instances is called a knowledge base. To illustrate some of the salient distinctions here, consider the cell line called A549, which is established from a human lung adenocarcinoma sample. There is no mention of A549 in the Cell Ontology. However, A549 is present in the EBI Experimental Factor Ontology as a subclass of the “Homo sapiens cell line” class. Presumably this is because A549 is a class of cells that are widely used experimentally, and this cell line constitutes a concept deserving of mapping in the universe of experimental factors. In the universe of concepts related to cell structure and function <em>per se</em>, A549 is an individual that can be characterized through possession of or lack of properties enumerated in Cell Ontology, but it is not deserving of inclusion in that ontology.</p>
</div>
<div id="illustration-in-a-single-cell-rna-seq-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#illustration-in-a-single-cell-rna-seq-dataset" class="anchor"></a>Illustration in a single-cell RNA-seq dataset</h2>
<p>The 10X Genomics corporation has distributed a dataset on results of sequencing 10000 PBMC from a healthy donor . Subsets of the data are used in tutorials for the Seurat analytical suite (<span class="citation">Butler et al. (2018)</span>).</p>
<div id="labeling-pbmc-in-the-seurat-tutorial" class="section level3">
<h3 class="hasAnchor">
<a href="#labeling-pbmc-in-the-seurat-tutorial" class="anchor"></a>Labeling PBMC in the Seurat tutorial</h3>
<p>One result of the tutorial analysis of the 3000 cell subset is a table of cell types and expression-based markers of cell identity. The first three columns of the table below are from concluding material in the Seurat tutorial; the remaining columns are created by “manual” matching between the Seurat terms and terms found in Cell Ontology.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="va">stab</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/seur3kTab.html">seur3kTab</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="right">grp</th>
<th align="left">markers</th>
<th align="left">seurTutType</th>
<th align="left">formal</th>
<th align="left">tag</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">0</td>
<td align="left">IL7R</td>
<td align="left">CD4 T cells</td>
<td align="left">CD4-positive helper T cell</td>
<td align="left">CL:0000492</td>
</tr>
<tr class="even">
<td align="right">1</td>
<td align="left">CD14, LYZ</td>
<td align="left">CD14+ Monocytes</td>
<td align="left">CD14-positive monocyte</td>
<td align="left">CL:0001054</td>
</tr>
<tr class="odd">
<td align="right">2</td>
<td align="left">MS4A1</td>
<td align="left">B cells</td>
<td align="left">B cell</td>
<td align="left">CL:0000236</td>
</tr>
<tr class="even">
<td align="right">3</td>
<td align="left">CD8A</td>
<td align="left">CD8 T cells</td>
<td align="left">CD8-positive, alpha-beta T cell</td>
<td align="left">CL:0000625</td>
</tr>
<tr class="odd">
<td align="right">4</td>
<td align="left">FCGR3A, MS4A7</td>
<td align="left">FCGR3A+ Monocytes</td>
<td align="left">monocyte</td>
<td align="left">CL:0000576</td>
</tr>
<tr class="even">
<td align="right">5</td>
<td align="left">GNLY, NKG7</td>
<td align="left">NK cells</td>
<td align="left">natural killer cell</td>
<td align="left">CL:0000623</td>
</tr>
<tr class="odd">
<td align="right">6</td>
<td align="left">FCER1A, CST3</td>
<td align="left">Dendritic Cells</td>
<td align="left">dendritic cell</td>
<td align="left">CL:0000451</td>
</tr>
<tr class="even">
<td align="right">7</td>
<td align="left">PPBP</td>
<td align="left">Megakaryocytes</td>
<td align="left">megakaryocyte</td>
<td align="left">CL:0000556</td>
</tr>
</tbody>
</table>
</div>
<div id="relationships-asserted-in-the-cell-ontology" class="section level3">
<h3 class="hasAnchor">
<a href="#relationships-asserted-in-the-cell-ontology" class="anchor"></a>Relationships asserted in the Cell Ontology</h3>
<p>Given the informally selected tags in the table above, we can sketch the Cell Ontology graph connecting the associated cell types. The ontoProc package adds functionality to ontologyPlot with <code>make_graphNEL_from_ontology_plot</code>. This allows use of all Rgraphviz and igraph visualization facilities for graphs derived from ontology structures.</p>
<p>Here we display the PBMC cell sets reported in the Seurat tutorial.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ontoProc</span><span class="op">)</span>
<span class="va">cl</span> <span class="op">=</span> <span class="fu"><a href="../reference/getCellOnto.html">getCellOnto</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="../reference/onto_plot2.html">onto_plot2</a></span><span class="op">(</span><span class="va">cl</span>, <span class="va">stab</span><span class="op">$</span><span class="va">tag</span><span class="op">)</span></code></pre></div>
<p><img src="ontoProc_files/figure-html/lklklk-1.png" width="700"></p>
</div>
<div id="molecular-features-asserted-in-the-cell-ontology" class="section level3">
<h3 class="hasAnchor">
<a href="#molecular-features-asserted-in-the-cell-ontology" class="anchor"></a>Molecular features asserted in the Cell Ontology</h3>
<p>The <code>CLfeats</code> function traces relationships and properties from a given Cell Ontology class. Briefly, each class can assert that it is the <code>intersection_of</code> other classes, and <code>has_part</code>, <code>lacks_part</code>, <code>has_plasma_membrane_part</code>, <code>lacks_plasma_membrane_part</code> can be asserted as relationships holding between cell type instances and cell components. The components are often cross-referenced to Protein Ontology or Gene Ontology. When the Protein Ontology component has a synonym for which an HGNC symbol is provided, that symbol is retrieved by <code>CLfeats</code>. Here we obtain the listing for a mature CD1a-positive dermal dendritic cell.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressMessages</a></span><span class="op">(</span><span class="op">{</span>
<span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="fu"><a href="../reference/CLfeats.html">CLfeats</a></span><span class="op">(</span><span class="va">cl</span>, <span class="st">"CL:0002531"</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span></code></pre></div>
<table class="table">
<colgroup>
<col width="8%">
<col width="10%">
<col width="7%">
<col width="39%">
<col width="5%">
<col width="28%">
</colgroup>
<thead><tr class="header">
<th align="left">tag</th>
<th align="left">prtag</th>
<th align="left">cond</th>
<th align="left">entity</th>
<th align="left">SYMBOL</th>
<th align="left">name</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">CL:0000990</td>
<td align="left">PR:000001013</td>
<td align="left">hiPMAmt</td>
<td align="left">integrin alpha-X</td>
<td align="left">ITGAX</td>
<td align="left">conventional dendritic cell</td>
</tr>
<tr class="even">
<td align="left">CL:0000451</td>
<td align="left">PR:000001002</td>
<td align="left">lacksPMP</td>
<td align="left">CD19 molecule</td>
<td align="left">CD19</td>
<td align="left">dendritic cell</td>
</tr>
<tr class="odd">
<td align="left">CL:0000451</td>
<td align="left">PR:000001003</td>
<td align="left">lacksPMP</td>
<td align="left">CD34 molecule</td>
<td align="left">CD34</td>
<td align="left">dendritic cell</td>
</tr>
<tr class="even">
<td align="left">CL:0000451</td>
<td align="left">PR:000001020</td>
<td align="left">lacksPMP</td>
<td align="left">CD3 epsilon</td>
<td align="left">CD3E</td>
<td align="left">dendritic cell</td>
</tr>
<tr class="odd">
<td align="left">CL:0000451</td>
<td align="left">PR:000001024</td>
<td align="left">lacksPMP</td>
<td align="left">neural cell adhesion molecule 1</td>
<td align="left">NCAM1</td>
<td align="left">dendritic cell</td>
</tr>
<tr class="even">
<td align="left">CL:0000451</td>
<td align="left">PR:000001289</td>
<td align="left">lacksPMP</td>
<td align="left">membrane-spanning 4-domains subfamily A member 1</td>
<td align="left">MS4A1</td>
<td align="left">dendritic cell</td>
</tr>
<tr class="odd">
<td align="left">CL:0000451</td>
<td align="left"><a href="GO:0042613" class="uri">GO:0042613</a></td>
<td align="left">hasPart</td>
<td align="left">MHC class II protein complex</td>
<td align="left">NA</td>
<td align="left">dendritic cell</td>
</tr>
<tr class="even">
<td align="left">CL:0000841</td>
<td align="left"><a href="GO:0042613" class="uri">GO:0042613</a></td>
<td align="left">hiPMAmt</td>
<td align="left">MHC class II protein complex</td>
<td align="left">NA</td>
<td align="left">mature conventional dendritic cell</td>
</tr>
<tr class="odd">
<td align="left">CL:0000841</td>
<td align="left">PR:000001412</td>
<td align="left">hiPMAmt</td>
<td align="left">CD86 molecule</td>
<td align="left">CD86</td>
<td align="left">mature conventional dendritic cell</td>
</tr>
<tr class="even">
<td align="left">CL:0000841</td>
<td align="left">PR:000001438</td>
<td align="left">hiPMAmt</td>
<td align="left">CD80 molecule</td>
<td align="left">CD80</td>
<td align="left">mature conventional dendritic cell</td>
</tr>
<tr class="odd">
<td align="left">CL:0000000</td>
<td align="left"><a href="GO:0005634" class="uri">GO:0005634</a></td>
<td align="left">hasPart</td>
<td align="left">nucleus</td>
<td align="left">NA</td>
<td align="left">cell</td>
</tr>
<tr class="even">
<td align="left">CL:0000000</td>
<td align="left"><a href="GO:0005634" class="uri">GO:0005634</a></td>
<td align="left">hasPart</td>
<td align="left">nucleus</td>
<td align="left">NA</td>
<td align="left">cell</td>
</tr>
</tbody>
</table>
<p>The <code>ctmarks</code> function starts a shiny app that generates tables of this sort for selected cell types.</p>
<div class="figure">
<img src="ctmarks.png" alt="ctmarks snapshot"><p class="caption">ctmarks snapshot</p>
</div>
</div>
<div id="mapping-from-gene-presencerole-to-cell-type" class="section level3">
<h3 class="hasAnchor">
<a href="#mapping-from-gene-presencerole-to-cell-type" class="anchor"></a>Mapping from gene ‘presence/role’ to cell type</h3>
<p>The <code>sym2CellOnto</code> function helps find mention of given gene symbols in properties or parts of cell types.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="va">sdf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="../reference/sym2CellOnto.html">sym2CellOnto</a></span><span class="op">(</span><span class="st">"ITGAM"</span>, <span class="va">cl</span>, <span class="va">pr</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<table class="table">
<colgroup>
<col width="6%">
<col width="9%">
<col width="11%">
<col width="73%">
</colgroup>
<thead><tr class="header">
<th align="left">sym</th>
<th align="left">cond</th>
<th align="left">cl</th>
<th align="left">type</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ITGAM</td>
<td align="left">lacksPMP</td>
<td align="left">CL:0000037</td>
<td align="left">hematopoietic stem cell</td>
</tr>
<tr class="even">
<td align="left">ITGAM</td>
<td align="left">lacksPMP</td>
<td align="left">CL:0000547</td>
<td align="left">proerythroblast</td>
</tr>
<tr class="odd">
<td align="left">ITGAM</td>
<td align="left">lacksPMP</td>
<td align="left">CL:0000553</td>
<td align="left">megakaryocyte progenitor cell</td>
</tr>
<tr class="even">
<td align="left">ITGAM</td>
<td align="left">lacksPMP</td>
<td align="left">CL:0000558</td>
<td align="left">reticulocyte</td>
</tr>
<tr class="odd">
<td align="left">ITGAM</td>
<td align="left">lacksPMP</td>
<td align="left">CL:0000611</td>
<td align="left">eosinophil progenitor cell</td>
</tr>
<tr class="even">
<td align="left">ITGAM</td>
<td align="left">lacksPMP</td>
<td align="left">CL:0000613</td>
<td align="left">basophil progenitor cell</td>
</tr>
<tr class="odd">
<td align="left">ITGAM</td>
<td align="left">lacksPMP</td>
<td align="left">CL:0000765</td>
<td align="left">erythroblast</td>
</tr>
<tr class="even">
<td align="left">ITGAM</td>
<td align="left">lacksPMP</td>
<td align="left">CL:0000831</td>
<td align="left">mast cell progenitor</td>
</tr>
<tr class="odd">
<td align="left">ITGAM</td>
<td align="left">lacksPMP</td>
<td align="left">CL:0000836</td>
<td align="left">promyelocyte</td>
</tr>
<tr class="even">
<td align="left">ITGAM</td>
<td align="left">lacksPMP</td>
<td align="left">CL:0000837</td>
<td align="left">hematopoietic multipotent progenitor cell</td>
</tr>
<tr class="odd">
<td align="left">ITGAM</td>
<td align="left">lacksPMP</td>
<td align="left">CL:0000872</td>
<td align="left">splenic marginal zone macrophage</td>
</tr>
<tr class="even">
<td align="left">ITGAM</td>
<td align="left">lacksPMP</td>
<td align="left">CL:0000941</td>
<td align="left">thymic conventional dendritic cell</td>
</tr>
<tr class="odd">
<td align="left">ITGAM</td>
<td align="left">lacksPMP</td>
<td align="left">CL:0000942</td>
<td align="left">thymic plasmacytoid dendritic cell</td>
</tr>
<tr class="even">
<td align="left">ITGAM</td>
<td align="left">lacksPMP</td>
<td align="left">CL:0000989</td>
<td align="left">CD11c-low plasmacytoid dendritic cell</td>
</tr>
<tr class="odd">
<td align="left">ITGAM</td>
<td align="left">lacksPMP</td>
<td align="left">CL:0000998</td>
<td align="left">CD8_alpha-negative CD11b-negative dendritic cell</td>
</tr>
<tr class="even">
<td align="left">ITGAM</td>
<td align="left">lacksPMP</td>
<td align="left">CL:0001000</td>
<td align="left">CD8_alpha-positive CD11b-negative dendritic cell</td>
</tr>
<tr class="odd">
<td align="left">ITGAM</td>
<td align="left">lacksPMP</td>
<td align="left">CL:0001029</td>
<td align="left">common dendritic progenitor</td>
</tr>
<tr class="even">
<td align="left">ITGAM</td>
<td align="left">lacksPMP</td>
<td align="left">CL:0001060</td>
<td align="left">hematopoietic oligopotent progenitor cell, lineage-negative</td>
</tr>
<tr class="odd">
<td align="left">ITGAM</td>
<td align="left">lacksPMP</td>
<td align="left">CL:0001066</td>
<td align="left">erythroid progenitor cell, mammalian</td>
</tr>
<tr class="even">
<td align="left">ITGAM</td>
<td align="left">lacksPMP</td>
<td align="left">CL:0002010</td>
<td align="left">pre-conventional dendritic cell</td>
</tr>
<tr class="odd">
<td align="left">ITGAM</td>
<td align="left">lacksPMP</td>
<td align="left">CL:0002089</td>
<td align="left">group 2 innate lymphoid cell, mouse</td>
</tr>
<tr class="even">
<td align="left">ITGAM</td>
<td align="left">lacksPMP</td>
<td align="left">CL:0002679</td>
<td align="left">natural helper lymphocyte</td>
</tr>
<tr class="odd">
<td align="left">ITGAM</td>
<td align="left">hiPMAmt</td>
<td align="left">CL:0000581</td>
<td align="left">peritoneal macrophage</td>
</tr>
<tr class="even">
<td align="left">ITGAM</td>
<td align="left">hiPMAmt</td>
<td align="left">CL:0002347</td>
<td align="left">CD27-high, CD11b-high natural killer cell, mouse</td>
</tr>
<tr class="odd">
<td align="left">ITGAM</td>
<td align="left">hiPMAmt</td>
<td align="left">CL:0002348</td>
<td align="left">CD27-low, CD11b-high natural killer cell, mouse</td>
</tr>
<tr class="even">
<td align="left">ITGAM</td>
<td align="left">hiPMAmt</td>
<td align="left">CL:0002474</td>
<td align="left">lymphoid MHC-II-negative classical monocyte</td>
</tr>
<tr class="odd">
<td align="left">ITGAM</td>
<td align="left">hiPMAmt</td>
<td align="left">CL:0002475</td>
<td align="left">lymphoid MHC-II-negative non-classical monocyte</td>
</tr>
<tr class="even">
<td align="left">ITGAM</td>
<td align="left">hiPMAmt</td>
<td align="left">CL:0002505</td>
<td align="left">liver CD103-negative dendritic cell</td>
</tr>
<tr class="odd">
<td align="left">ITGAM</td>
<td align="left">hiPMAmt</td>
<td align="left">CL:0002510</td>
<td align="left">CD103-negative, langerin-positive lymph node dendritic cell</td>
</tr>
<tr class="even">
<td align="left">ITGAM</td>
<td align="left">hiPMAmt</td>
<td align="left">CL:0002512</td>
<td align="left">CD11b-high, CD103-negative, langerin-negative lymph node dendritic cell</td>
</tr>
<tr class="odd">
<td align="left">ITGAM</td>
<td align="left">loPMAmt</td>
<td align="left">CL:0000091</td>
<td align="left">Kupffer cell</td>
</tr>
<tr class="even">
<td align="left">ITGAM</td>
<td align="left">loPMAmt</td>
<td align="left">CL:0000583</td>
<td align="left">alveolar macrophage</td>
</tr>
<tr class="odd">
<td align="left">ITGAM</td>
<td align="left">loPMAmt</td>
<td align="left">CL:0000868</td>
<td align="left">lymph node macrophage</td>
</tr>
<tr class="even">
<td align="left">ITGAM</td>
<td align="left">loPMAmt</td>
<td align="left">CL:0000874</td>
<td align="left">splenic red pulp macrophage</td>
</tr>
<tr class="odd">
<td align="left">ITGAM</td>
<td align="left">loPMAmt</td>
<td align="left">CL:0002345</td>
<td align="left">CD27-low, CD11b-low immature natural killer cell, mouse</td>
</tr>
<tr class="even">
<td align="left">ITGAM</td>
<td align="left">loPMAmt</td>
<td align="left">CL:0002349</td>
<td align="left">CD27-high, CD11b-low natural killer cell, mouse</td>
</tr>
<tr class="odd">
<td align="left">ITGAM</td>
<td align="left">loPMAmt</td>
<td align="left">CL:0002506</td>
<td align="left">liver CD103-positive dendritic cell</td>
</tr>
<tr class="even">
<td align="left">ITGAM</td>
<td align="left">loPMAmt</td>
<td align="left">CL:0002509</td>
<td align="left">CD103-positive, langerin-positive lymph node dendritic cell</td>
</tr>
<tr class="odd">
<td align="left">ITGAM</td>
<td align="left">loPMAmt</td>
<td align="left">CL:0002511</td>
<td align="left">CD11b-low, CD103-negative, langerin-negative lymph node dendritic cell</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">sdf</span><span class="op">$</span><span class="va">cond</span><span class="op">)</span></code></pre></div>
<pre><code>## 
##  hiPMAmt lacksPMP  loPMAmt 
##        8       22        9</code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="../reference/sym2CellOnto.html">sym2CellOnto</a></span><span class="op">(</span><span class="st">"FOXP3"</span>, <span class="va">cl</span>, <span class="va">pr</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<table class="table">
<colgroup>
<col width="7%">
<col width="9%">
<col width="13%">
<col width="69%">
</colgroup>
<thead><tr class="header">
<th align="left">sym</th>
<th align="left">cond</th>
<th align="left">cl</th>
<th align="left">type</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">FOXP3</td>
<td align="left">hasPart</td>
<td align="left">CL:0000902</td>
<td align="left">induced T-regulatory cell</td>
</tr>
<tr class="even">
<td align="left">FOXP3</td>
<td align="left">hasPart</td>
<td align="left">CL:0000903</td>
<td align="left">natural T-regulatory cell</td>
</tr>
<tr class="odd">
<td align="left">FOXP3</td>
<td align="left">hasPart</td>
<td align="left">CL:0000919</td>
<td align="left">CD8-positive, CD25-positive, alpha-beta regulatory T cell</td>
</tr>
<tr class="even">
<td align="left">FOXP3</td>
<td align="left">hasPart</td>
<td align="left">CL:0000920</td>
<td align="left">CD8-positive, CD28-negative, alpha-beta regulatory T cell</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="adding-terms-to-ontology_index-structures-to-extend-cell-ontology" class="section level2">
<h2 class="hasAnchor">
<a href="#adding-terms-to-ontology_index-structures-to-extend-cell-ontology" class="anchor"></a>Adding terms to ontology_index structures to ‘extend’ Cell Ontology</h2>
<p>The task of extending an ontology is partly bureaucratic in nature and depends on a collection of endorsements and updates to centralized information structures. In order to permit experimentation with interfaces and new content that may be quite speculative, we include an approach to combining new ontology ‘terms’ of structure similar to those endorsed in Cell Ontology, to ontologyIndex-based <code>ontology_index</code> instances.</p>
<div id="use-case-a-set-of-cell-types-defined-by-diagonal-expression" class="section level3">
<h3 class="hasAnchor">
<a href="#use-case-a-set-of-cell-types-defined-by-diagonal-expression" class="anchor"></a>Use case: a set of cell types defined by “diagonal expression”</h3>
<p>For a demonstration, we consider the discussion in <span class="citation">Bakken et al. (2017)</span>, of a ‘diagonal’ expression pattern defining a group of novel cell types. A set of genes is identified and cells are distinguised by expressing exactly one gene from the set.</p>
<div class="figure">
<img src="usecaseCyclic.png" alt="Diagonal expression pattern."><p class="caption">Diagonal expression pattern.</p>
</div>
<p>The necessary information is collected in a vector. The vector is the set of genes, the name of element i is the tag to be associated with the type of cell that expresses gene i and does not express any other gene in the set.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sigels</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CL:X01"</span><span class="op">=</span><span class="st">"GRIK3"</span>, <span class="st">"CL:X02"</span><span class="op">=</span><span class="st">"NTNG1"</span>, <span class="st">"CL:X03"</span><span class="op">=</span><span class="st">"BAGE2"</span>,
             <span class="st">"CL:X04"</span><span class="op">=</span><span class="st">"MC4R"</span>, <span class="st">"CL:X05"</span><span class="op">=</span><span class="st">"PAX6"</span>, <span class="st">"CL:X06"</span><span class="op">=</span><span class="st">"TSPAN12"</span>, 
             <span class="st">"CL:X07"</span><span class="op">=</span><span class="st">"hSHISA8"</span>, <span class="st">"CL:X08"</span><span class="op">=</span><span class="st">"SNCG"</span>, <span class="st">"CL:X09"</span><span class="op">=</span><span class="st">"ARHGEF28"</span>, 
             <span class="st">"CL:X10"</span><span class="op">=</span><span class="st">"EGF"</span><span class="op">)</span></code></pre></div>
</div>
<div id="a-data-frame-defining-the-cell-types-and-their-properties" class="section level3">
<h3 class="hasAnchor">
<a href="#a-data-frame-defining-the-cell-types-and-their-properties" class="anchor"></a>A data.frame defining the cell types and their properties</h3>
<p>The <code>cyclicSigset</code> function produces a data.frame instance connecting cell types with the genes expressed or unexpressed.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cs</span> <span class="op">=</span> <span class="fu"><a href="../reference/cyclicSigset.html">cyclicSigset</a></span><span class="op">(</span><span class="va">sigels</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">cs</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 100   3</code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cs</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">9</span><span class="op">:</span><span class="fl">13</span><span class="op">)</span>,<span class="op">]</span></code></pre></div>
<pre><code>##        gene   type     cond
## 1  ARHGEF28 CL:X09   hasExp
## 2     GRIK3 CL:X09 lacksExp
## 3     NTNG1 CL:X09 lacksExp
## 4     BAGE2 CL:X09 lacksExp
## 5      MC4R CL:X09 lacksExp
## 9      SNCG CL:X09 lacksExp
## 10      EGF CL:X09 lacksExp
## 11    BAGE2 CL:X03   hasExp
## 12    GRIK3 CL:X03 lacksExp
## 13    NTNG1 CL:X03 lacksExp</code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">cs</span><span class="op">$</span><span class="va">cond</span><span class="op">)</span></code></pre></div>
<pre><code>## 
##   hasExp lacksExp 
##       10       90</code></pre>
<p>It is expected that a tabular layout like this will suffice to handle general situations of cell type definition.</p>
</div>
<div id="translating-the-data-frame-elements-to-obo-term-instances" class="section level3">
<h3 class="hasAnchor">
<a href="#translating-the-data-frame-elements-to-obo-term-instances" class="anchor"></a>Translating the data.frame elements to OBO Term instances</h3>
<p>The most complicated aspect of novel OBO term construction is the proper specifications of relationships with existing ontology components. A prolog that is mostly shared by all terms is generated programmatically for the diagonal pattern task.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"> <span class="va">makeIntnProlog</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">id</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span>
 <span class="co"># make type-specific prologs as key-value pairs</span>
     <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>  
       <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html">sprintf</a></span><span class="op">(</span><span class="st">"id: %s"</span>, <span class="va">id</span><span class="op">)</span>,
       <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html">sprintf</a></span><span class="op">(</span><span class="st">"name: %s-expressing cortical layer 1 interneuron, human"</span>, <span class="va">...</span><span class="op">)</span>,
       <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html">sprintf</a></span><span class="op">(</span><span class="st">"def: '%s-expressing cortical layer 1 interneuron, human described via RNA-seq observations' [PMID 29322913]"</span>, <span class="va">...</span><span class="op">)</span>,
       <span class="st">"is_a: CL:0000099 ! interneuron"</span>,
       <span class="st">"intersection_of: CL:0000099 ! interneuron"</span><span class="op">)</span>
 <span class="op">}</span></code></pre></div>
<p>The <code>ldfToTerms</code> API uses this to create a set of strings that can be parsed as a term.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pmap</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"hasExp"</span><span class="op">=</span><span class="st">"has_expression_of"</span>, lacksExp<span class="op">=</span><span class="st">"lacks_expression_of"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">tms</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ldfToTerms.html">ldfToTerms</a></span><span class="op">(</span><span class="va">cs</span>, <span class="va">pmap</span>, <span class="va">sigels</span>, <span class="va">makeIntnProlog</span><span class="op">)</span><span class="op">)</span>, <span class="fl">20</span><span class="op">)</span></code></pre></div>
<pre><code>##  [1] "[Term]"                                                                                                        
##  [2] "id: CL:X01"                                                                                                    
##  [3] "name: GRIK3-expressing cortical layer 1 interneuron, human"                                                    
##  [4] "def: 'GRIK3-expressing cortical layer 1 interneuron, human described via RNA-seq observations' [PMID 29322913]"
##  [5] "is_a: CL:0000099 ! interneuron"                                                                                
##  [6] "intersection_of: CL:0000099 ! interneuron"                                                                     
##  [7] "has_expression_of: PR:000008242 ! GRIK3"                                                                       
##  [8] "lacks_expression_of: PR:000011467 ! NTNG1"                                                                     
##  [9] "lacks_expression_of: PR:000004625 ! BAGE2"                                                                     
## [10] "lacks_expression_of: PR:000001237 ! MC4R"                                                                      
## [11] "lacks_expression_of: PR:000012318 ! PAX6"                                                                      
## [12] "lacks_expression_of: PR:000016738 ! TSPAN12"                                                                   
## [13] "lacks_expression_of: PR:B8ZZ34 ! hSHISA8"                                                                      
## [14] "lacks_expression_of: PR:000015325 ! SNCG"                                                                      
## [15] "lacks_expression_of: PR:000013942 ! ARHGEF28"                                                                  
## [16] "lacks_expression_of: PR:000006928 ! EGF"                                                                       
## [17] "[Term]"                                                                                                        
## [18] "id: CL:X02"                                                                                                    
## [19] "name: NTNG1-expressing cortical layer 1 interneuron, human"                                                    
## [20] "def: 'NTNG1-expressing cortical layer 1 interneuron, human described via RNA-seq observations' [PMID 29322913]"</code></pre>
<p>The content in tms can then be appended to the content of the Cell Ontology cl.obo as text for import with <code><a href="https://rdrr.io/pkg/ontologyIndex/man/get_ontology.html">ontologyIndex::get_OBO</a></code>.</p>
</div>
</div>
<div id="subsetting-singler-resources-using-ontological-mapping" class="section level2">
<h2 class="hasAnchor">
<a href="#subsetting-singler-resources-using-ontological-mapping" class="anchor"></a>Subsetting SingleR resources using ontological mapping</h2>
<div id="a-data-frame-mapping-from-informal-to-formal-terms" class="section level3">
<h3 class="hasAnchor">
<a href="#a-data-frame-mapping-from-informal-to-formal-terms" class="anchor"></a>A data.frame mapping from informal to formal terms</h3>
<p>Aaron Lun has produced a mapping from informal terms used in the Human Primary Cell Atlas to Cell Ontology tags. We provisionally include a copy of this mapping in ontoProc:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hpca_map</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/hpca.csv"</span>, package<span class="op">=</span><span class="st">"ontoProc"</span><span class="op">)</span>, strings<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">hpca_map</span><span class="op">)</span></code></pre></div>
<pre><code>##                          uncontrolled controlled
## 1        DC:monocyte-derived:immature CL:0000840
## 2      DC:monocyte-derived:Galectin-1 CL:0000451
## 3             DC:monocyte-derived:LPS CL:0000451
## 4                 DC:monocyte-derived CL:0000451
## 5 Smooth_muscle_cells:bronchial:vit_D CL:0002598
## 6       Smooth_muscle_cells:bronchial CL:0002598</code></pre>
<p>We will rename columns of this map for convenience of our <code>bind_formal_tags</code> method.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">hpca_map</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"informal"</span>, <span class="st">"formal"</span><span class="op">)</span>  <span class="co"># obligatory for now</span></code></pre></div>
</div>
<div id="binding-formal-tags-to-the-hpca-data" class="section level3">
<h3 class="hasAnchor">
<a href="#binding-formal-tags-to-the-hpca-data" class="anchor"></a>Binding formal tags to the HPCA data</h3>
<p>I am turning this code off for now because there is no standard approach to getting the mapping from the SummarizedExperment yet. When SingleR merges the ‘standardized’ branch, this will come back.</p>
<p>Let’s retrieve the HPCA data from SingleR:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></code></pre></div>
<pre><code>## Loading required package: SummarizedExperiment</code></pre>
<pre><code>## Loading required package: MatrixGenerics</code></pre>
<pre><code>## Loading required package: matrixStats</code></pre>
<pre><code>## 
## Attaching package: 'MatrixGenerics'</code></pre>
<pre><code>## The following objects are masked from 'package:matrixStats':
## 
##     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
##     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
##     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
##     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
##     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
##     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
##     colWeightedMeans, colWeightedMedians, colWeightedSds,
##     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
##     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
##     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
##     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
##     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
##     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
##     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
##     rowWeightedSds, rowWeightedVars</code></pre>
<pre><code>## Loading required package: GenomicRanges</code></pre>
<pre><code>## Loading required package: stats4</code></pre>
<pre><code>## Loading required package: BiocGenerics</code></pre>
<pre><code>## Loading required package: parallel</code></pre>
<pre><code>## 
## Attaching package: 'BiocGenerics'</code></pre>
<pre><code>## The following objects are masked from 'package:parallel':
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>## The following objects are masked from 'package:stats':
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     anyDuplicated, append, as.data.frame, basename, cbind, colnames,
##     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
##     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
##     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
##     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
##     union, unique, unsplit, which.max, which.min</code></pre>
<pre><code>## Loading required package: S4Vectors</code></pre>
<pre><code>## 
## Attaching package: 'S4Vectors'</code></pre>
<pre><code>## The following object is masked from 'package:base':
## 
##     expand.grid</code></pre>
<pre><code>## Loading required package: IRanges</code></pre>
<pre><code>## Loading required package: GenomeInfoDb</code></pre>
<pre><code>## Loading required package: Biobase</code></pre>
<pre><code>## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with
##     'browseVignettes()'. To cite Bioconductor, see
##     'citation("Biobase")', and for packages 'citation("pkgname")'.</code></pre>
<pre><code>## 
## Attaching package: 'Biobase'</code></pre>
<pre><code>## The following object is masked from 'package:MatrixGenerics':
## 
##     rowMedians</code></pre>
<pre><code>## The following objects are masked from 'package:matrixStats':
## 
##     anyMissing, rowMedians</code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/celldex">celldex</a></span><span class="op">)</span>
<span class="va">hpca_sce</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/celldex/man/HumanPrimaryCellAtlasData.html">HumanPrimaryCellAtlasData</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## snapshotDate(): 2020-10-27</code></pre>
<pre><code>## see ?celldex and browseVignettes('celldex') for documentation</code></pre>
<pre><code>## loading from cache</code></pre>
<pre><code>## see ?celldex and browseVignettes('celldex') for documentation</code></pre>
<pre><code>## loading from cache</code></pre>
<p>Now bind the formal tags:</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hpca_sce</span> <span class="op">=</span> <span class="fu"><a href="../reference/bind_formal_tags.html">bind_formal_tags</a></span><span class="op">(</span><span class="va">hpca_sce</span>, <span class="st">"label.fine"</span>, <span class="va">hpca_map</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">hpca_sce</span><span class="op">$</span><span class="va">label.ont</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 66</code></pre>
<p>We don’t check for failed mappings:</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">xx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">hpca_sce</span><span class="op">$</span><span class="va">label.ont</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 2</code></pre>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">xx</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">hpca_sce</span><span class="op">)</span><span class="op">[</span><span class="va">xx</span>,<span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>## DataFrame with 2 rows and 3 columns
##            label.main             label.fine   label.ont
##           &lt;character&gt;            &lt;character&gt; &lt;character&gt;
## GSM154081     T_cells T_cell:CCR10+CLA+1,2..          NA
## GSM154084     T_cells T_cell:CCR10-CLA+1,2..          NA</code></pre>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">hpca_sce</span><span class="op">$</span><span class="va">label.ont</span> <span class="op">==</span> <span class="st">""</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="co"># iPS and BM</span></code></pre></div>
<pre><code>## [1] 50</code></pre>
</div>
<div id="subsetting-using-the-class-hierarchy-of-cell-ontology" class="section level3">
<h3 class="hasAnchor">
<a href="#subsetting-using-the-class-hierarchy-of-cell-ontology" class="anchor"></a>Subsetting using the class hierarchy of Cell Ontology</h3>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cell_onto</span> <span class="op">=</span> <span class="fu">ontoProc</span><span class="fu">::</span><span class="fu"><a href="../reference/getCellOnto.html">getCellOnto</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">hpca_mono</span> <span class="op">=</span> <span class="fu"><a href="../reference/subset_descendants.html">subset_descendants</a></span><span class="op">(</span> <span class="va">hpca_sce</span>, <span class="va">cell_onto</span>, <span class="st">"^monocyte$"</span> <span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">hpca_mono</span><span class="op">$</span><span class="va">label.fine</span><span class="op">)</span></code></pre></div>
<pre><code>## 
##                          Monocyte             Monocyte:anti-FcgRIIB 
##                                27                                 2 
##                    Monocyte:CD14+                    Monocyte:CD16- 
##                                 3                                 7 
##                    Monocyte:CD16+                    Monocyte:CXCL4 
##                                 6                                 2 
##   Monocyte:F._tularensis_novicida           Monocyte:leukotriene_D4 
##                                 6                                 4 
##                     Monocyte:MCSF Monocyte:S._typhimurium_flagellin 
##                                 2                                 1</code></pre>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">hpca_mono</span><span class="op">$</span><span class="va">label.ont</span><span class="op">)</span> <span class="co"># not much diversity</span></code></pre></div>
<pre><code>## 
## CL:0000576 CL:0001054 
##         57          3</code></pre>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hpca_tcell</span> <span class="op">=</span> <span class="fu"><a href="../reference/subset_descendants.html">subset_descendants</a></span><span class="op">(</span> <span class="va">hpca_sce</span>, <span class="va">cell_onto</span>, <span class="st">"^T cell$"</span> <span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">hpca_tcell</span><span class="op">$</span><span class="va">label.fine</span><span class="op">)</span></code></pre></div>
<pre><code>## 
##                    T_cell:CD4+     T_cell:CD4+_central_memory 
##                             12                              5 
##    T_cell:CD4+_effector_memory              T_cell:CD4+_Naive 
##                              4                              6 
##                    T_cell:CD8+     T_cell:CD8+_Central_memory 
##                             16                              3 
##    T_cell:CD8+_effector_memory T_cell:CD8+_effector_memory_RA 
##                              4                              4 
##              T_cell:CD8+_naive                T_cell:effector 
##                              4                              4 
##             T_cell:gamma-delta              T_cell:Treg:Naive 
##                              2                              2</code></pre>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">hpca_tcell</span><span class="op">$</span><span class="va">label.ont</span><span class="op">)</span> <span class="co"># </span></code></pre></div>
<pre><code>## 
## CL:0000624 CL:0000625 CL:0000798 CL:0000895 CL:0000900 CL:0000904 CL:0000905 
##         12         16          2          6          4          5          4 
## CL:0000907 CL:0000911 CL:0000913 CL:0001062 CL:0002677 
##          3          4          4          4          2</code></pre>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">uu</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">hpca_tcell</span><span class="op">$</span><span class="va">label.ont</span><span class="op">)</span>
<span class="fu"><a href="../reference/onto_plot2.html">onto_plot2</a></span><span class="op">(</span><span class="va">cell_onto</span>, <span class="va">uu</span><span class="op">)</span></code></pre></div>
<p><img src="ontoProc_files/figure-html/dosub-1.png" width="700"></p>
</div>
</div>
</div>
<div id="related-tools" class="section level1">
<h1 class="hasAnchor">
<a href="#related-tools" class="anchor"></a>Related tools</h1>
<p>Inference on the identities of cells assayed in a single cell transcriptomics experiment can be performed using the Bioconductor <code>celaref</code> package. This package includes a number of reference data resources providing whole-transcriptome profiles of cells of known and unknown type. An approach to systematically structuring data on cell-type signatures, and conducting inference on cell types in new experiments, is provided in the <a href="https://github.com/kevinrue/Hancock">Hancock package</a>, under development.</p>
<p>A CRAN package that is very useful for R programming with ontologies is <code>ontologyIndex</code> <span class="citation">Westbury et al. (2015)</span>. This provides easily used functions for parsing ontologies in the OBO format and for performing basic queries on text fields and list structures.</p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Aevermann2018">
<p>Aevermann, Brian D., Mark Novotny, Trygve Bakken, Jeremy A. Miller, Alexander D. Diehl, David Osumi-Sutherland, Roger S. Lasken, Ed S. Lein, and Richard H. Scheuermann. 2018. “Cell type discovery using single-cell transcriptomics: Implications for ontological representation.” <em>Human Molecular Genetics</em> 27 (R1): R40–R47. <a href="https://doi.org/10.1093/hmg/ddy100">https://doi.org/10.1093/hmg/ddy100</a>.</p>
</div>
<div id="ref-Bakken2017">
<p>Bakken, Trygve, Lindsay Cowell, Brian D. Aevermann, Mark Novotny, Rebecca Hodge, Jeremy A. Miller, Alexandra Lee, et al. 2017. “Cell type discovery and representation in the era of high-content single cell phenotyping.” <em>BMC Bioinformatics</em> 18 (Suppl 17). <a href="https://doi.org/10.1186/s12859-017-1977-1">https://doi.org/10.1186/s12859-017-1977-1</a>.</p>
</div>
<div id="ref-butler">
<p>Butler, Andrew, Paul Hoffman, Peter Smibert, Efthymia Papalexi, and Rahul Satija. 2018. “Integrating Single-Cell Transcriptomic Data Across Different Conditions, Technologies, and Species.” <em>Nature Biotechnology</em>. <a href="https://doi.org/10.1038/nbt.4096">https://doi.org/10.1038/nbt.4096</a>.</p>
</div>
<div id="ref-Westbury2015">
<p>Westbury, Sarah K., Ernest Turro, Daniel Greene, Claire Lentaigne, Anne M. Kelly, Tadbir K. Bariana, Ilenia Simeoni, et al. 2015. “Human Phenotype Ontology Annotation and Cluster Analysis to Unravel Genetic Defects in 707 Cases with Unexplained Bleeding and Platelet Disorders.” <em>Genome Medicine</em> 7 (1): 36. <a href="https://doi.org/10.1186/s13073-015-0151-5">https://doi.org/10.1186/s13073-015-0151-5</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by VJ Carey.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
